Add blog post: "When Secure-by-Design Is Not Enough" (MSc thesis summary) (#51)
CHERIoT-Platform/CHERIoT-Platform.github.io: _posts/2026-02-22-secure-by-design-not-enough.md

A proof is only as strong as the model it is built on. But if the model accurately reflects the system of interest, formal verification allows us to move from “this seems safe” to “this property holds for all possible executions”.

In my thesis, I set out to apply this approach to some of the most privileged components in CHERIoT — and to do so at a level much closer to what actually runs on the hardware than is typical in software verification.

## What I verified

I focused on two of the most privileged components in the CHERIoT RTOS: the *unsealer* (part of the allocator) and the *switcher*.

> Both sit deep in the trusted computing base. Both manipulate capabilities directly. And both must uphold strict capability-safety guarantees for the entire system to remain sound.

Reviewer: Perhaps expand on "sit deep in the TCB" a bit: both of these components hold unsealing rights that allow them to see and manipulate things non-TCB components _must not_ be allowed to see or manipulate.  (Register spill areas & trusted stacks for the switcher, token object headers for the unsealer.)

"manipulate capabilities directly" is also true of a lot of non-TCB components in a CHERIoT firmware, but, as you say, what sets these apart is that their manipulations are part of the soundness of the system.

Perhaps the two above thoughts can be combined with something like...
```
Both sit deep in the trusted computing base.
As these components serve to define parts of the system's behavior,
they necessarily hold and manipulate capabilities sufficiently powerful to _break_ aspects of system soundness.
(The switcher especially is central to all aspects of the CIA triad.)
As such, both of these components must uphold...
```

### The unsealer

> The unsealer lives inside the allocator, which manages dynamic memory and provides use-after-free protection. In CHERIoT, sealed capabilities are used to construct type-safe, opaque abstractions (see the [introduction to sealed types](https://cheriot.org/sealing/compiler/2025/01/30/introducing-sealed-types.html)).

Reviewer 2: Can you use one line per sentence?  It makes review easier.

```suggestion
The unsealer is a special privileged library. In CHERIoT, sealed capabilities are used to construct type-safe, opaque abstractions (see the [introduction to sealed types](https://cheriot.org/sealing/compiler/2025/01/30/introducing-sealed-types.html)).
```

Reviewer: Perhaps "a special privileged library that works in concert with the RTOS loader and heap allocator", if more detail is desired?

At the ISA level, sealing combines a pointer-like capability with a sealing key. The result is a capability that can be passed around, but not dereferenced or modified. Unsealing reverses this, but only with the correct authority. As [David explains](https://cheriot.org/rtos/sealing/2025/11/06/sealing.html):

> “The `token_unseal` function unseals the object using the hardware sealing key and then compares the virtual otype in the header to the type of the permit-unseal capability passed as the key. If they match, it returns an unsealed capability to the object without the header, so the caller doesn’t ever have access to the header.”

> The `token_unseal` implementation does not rely solely on the user-provided unsealing key. It also accesses a hardware sealing key — effectively a master unsealing capability. That is acceptable because the allocator is trusted, but it is absolutely critical that this authority never leaks. Similarly, the header of the sealed object contains sensitive information that must never be exposed to untrusted code.

Reviewer 2: I don't understand 'That is acceptable because the allocator is trusted' here.  The allocator isn't involved.

Reviewer: Just s/allocator/unsealer/ ?

Author: Yeah, I got from the [cheriot-rtos architecture documentation](https://github.com/CHERIoT-Platform/cheriot-rtos/blob/main/docs/architecture.md) that the main components of the RTOS are switcher, allocator, scheduler and loader. So I wanted to situate the unsealer in this and got it into my head that the unsealer is part of the allocator.

Author: I'm just gonna change it to "the unsealer is trusted"

> The `token_unseal` implementation does not rely solely on the user-provided unsealing key. It also accesses a hardware sealing key — effectively a master unsealing capability. That is acceptable because the allocator is trusted, but it is absolutely critical that this authority never leaks. Similarly, the header of the sealed object contains sensitive information that must never be exposed to untrusted code.

Reviewer: I might say _cannot_ rely solely rather than "does not": it's implementing the virtualized sealing mechanism that we've designed / adopted / fallen back to because we didn't have enough architectural bits in capabilities.

> 

Reviewer 2: It would be good to explain that the token library is privileged, but is not invoked via a compartment transition and so must be defensively programmed to not leak state.  Effectively, it's implementing a small special-cased subset of switcher behaviour itself.

The properties I verified were therefore security properties:
- The hardware sealing capability must never be exposed.
- No capability derived from it may remain accessible after execution.
- The header of the sealed object must not leak to the caller.
- The only capability returned must be the correctly unsealed object — and only if the virtual object type matches.

### The switcher
While the allocator controls dynamic memory, the switcher controls execution itself. It is the most privileged component of the RTOS and is responsible for context switching between compartments and threads.

Its responsibilities include:
- Saving register state on interrupts,
- Invoking the scheduler with sealed references to the register save area and trusted stack,
- Restoring general-purpose and capability registers from the trusted stack,
- Installing a new program counter capability (PCC),
- Resuming execution of the selected context.

> 

Reviewer: It might be worth while spelling out a bit of "given this authority, what can go wrong".  The switcher can, from a "what's everything I can do with these capabilities" perspective...

- mutate (things reachable from) thread register contents (breaking integrity)
- copy (things reachable from) thread register contents between threads (breaking confidentiality)
- mutate compartment global state (again, integrity)
- copy compartment global state between compartments (again, confidentiality)
- refuse to make (certain) cross-compartment calls or refuse to abide (certain) scheduling calls (breaking availability)

- Strict isolation across execution contexts in the switcher.

Both sit at the heart of CHERIoT’s security model — and both operate at the lowest level of the system.

## The key idea - Binary-level verification on RTL

Most formal verification of low-level software operates at the source-code level — typically C or C++. Tools analyse the program and prove that certain properties hold at that level: that a variable is not zero, that a pointer does not escape its bounds, that a particular invariant is maintained.

But in CHERIoT, we care about the actual architectural state of the processor: the contents of the register file, the permissions attached to active capabilities, and whether authority is confined exactly as intended.

This creates an *abstraction gap*. Source code must still pass through compilers, linkers, and runtime layers before it executes. Even verification at the assembly level typically relies on an abstract model of the ISA semantics, not the actual hardware implementation. Compiler bugs, microarchitectural corner cases, or mismatches between specification and silicon remain outside the scope of the proof.

So I took a different approach.

> Instead of verifying source code against an abstract model of the ISA, I verified compiled binary code running directly on a cycle-accurate RTL model of the CHERIoT-Ibex processor.

Reviewer 2: It would be good to mention the prior work here that proved that the CHERIoT Ibex implements the ISA, so proofs done against the Ibex should be valid on any implementation of the ISA.

Concretely, I took:

> - The compiled binary of the relevant RTOS components,

Reviewer 2: Assembled maybe?  Neither of the components that you looked at are compiled, though they do use macros in assembly.

This eliminates the compilation gap: the model I reason about is much closer to what actually executes on the device.

It also makes the problem significantly harder.

At the RTL level, there are no variables or functions, only wires. The processor is a transition system: at each clock cycle, signals are either high or low. To prove a security property, I had to express it in terms of these signals across time — in the language of SystemVerilog and temporal logic.

From this, I built a framework that encodes the execution of specific binary instruction sequences and checks the desired capability-safety properties directly against the processor model.

## How this works in practice

At a high level, the problem had three parts:

1. Define what it means for a piece of binary code to execute.  
2. Define the security properties precisely.  
3. Prove that those properties hold for every possible execution.

Each of these required translating informal reasoning about code into precise statements over RTL signals.

### 1. Defining execution in Linear Temporal Logic

The first step was to formalise what it means for a sequence of instructions to execute on the processor.

> Since we are working at the RTL level, execution must be expressed in terms of pipeline signals. I chose the **writeback stage** as the “point of truth”. An instruction is considered executed when it reaches writeback without a fetch error. This fixes the final pipeline stage and avoids reasoning about partially executed instructions.

Reviewer: For my own edification: do you consider, say, mispredicted instructions arriving at writeback to be "executed but squashed and so of no effect" or something like that?  Similarly for instructions that may have faulted in earlier stages?
